rust/template/api.rs

```diff
@@ -0,0 +1,580 @@
+#![allow(non_snake_case, non_camel_case_types, non_upper_case_globals, unused_parens, non_shorthand_field_patterns, dead_code, overflowing_literals)]
+
+use differential_datalog::program::*;
+use differential_datalog::record;
+use differential_datalog::record::IntoRecord;
+
+use std::os::raw;
+use std::ptr;
+use std::ffi;
+use std::fs;
+use std::io::Write;
+use std::os::unix;
+use std::os::unix::io::{IntoRawFd, FromRawFd};
+use std::mem;
+use std::sync::{Mutex, Arc};
+use libc::size_t;
+
+use super::valmap::*;
+use super::update_handler::*;
+use super::*;
+
+pub struct HDDlog {
+    pub prog: Mutex<RunningProgram<Value>>,
+    pub update_handler: Box<dyn IMTUpdateHandler<Value>>,
+    pub db: Option<Arc<Mutex<ValMap>>>,
+    pub deltadb: Arc<Mutex<Option<DeltaMap>>>,
+    pub print_err: Option<extern "C" fn(msg: *const raw::c_char)>,
+    /* When set, all commands sent to the program are recorded in
+     * the specified `.dat` file so that they can be replayed later. */
+    pub replay_file: Option<Mutex<fs::File>>
+}
+
+impl HDDlog {
+    pub fn print_err(f: Option<extern "C" fn(msg: *const raw::c_char)>, msg: &str) {
+        match f {
+            None    => eprintln!("{}", msg),
+            Some(f) => f(ffi::CString::new(msg).unwrap().into_raw())
+        }
+    }
+
+    pub fn eprintln(&self, msg: &str)
+    {
+        Self::print_err(self.print_err, msg)
+    }
+}
+
+pub fn updcmd2upd(c: &record::UpdCmd) -> Result<Update<Value>, String>
+{
+    match c {
+        record::UpdCmd::Insert(rident, rec) => {
```

Comment:
I am guessing that if match fails you get automatically a Result that is not Ok.

Reply:
The pattern is exhaustive (and the compiler knows it)